dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java

```diff
@@ -1328,6 +1331,106 @@ private List<Event> inlineMemCpy(FunctionCall call) {
+        final Expression countExpr = expressions.makeCast(call.getArguments().get(3), types.getArchType());
+
+        if (!(countExpr instanceof IntLiteral countValue)) {
+            final String error = "Cannot handle memcpy_s with dynamic count argument: " + call;
+            throw new UnsupportedOperationException(error);
+        }
+        final int count = countValue.getValueAsInt();
+        if (!(destszExpr instanceof IntLiteral destszValue)) {
+            final String error = "Cannot handle memcpy_s with dynamic destsz argument: " + call;
+            throw new UnsupportedOperationException(error);
+        }
+        final int destsz = destszValue.getValueAsInt();
+
+        // Runtime checks
+        final Expression nullExpr = expressions.makeZero(types.getArchType());
+        final Expression destIsNull = expressions.makeEQ(dest, nullExpr);
+        final Expression srcIsNull = expressions.makeEQ(src, nullExpr);
+        // We assume RSIZE_MAX = 2^64-1
+        final Expression rsize_max = expressions.makeValue(BigInteger.ONE.shiftLeft(64).subtract(BigInteger.ONE), types.getArchType());
+        final Expression invalidDestsz = expressions.makeGT(destszExpr, rsize_max, false);
+        final Expression countGtMax = expressions.makeGT(countExpr, rsize_max, false);
+        final Expression countGtdestszExpr = expressions.makeGT(countExpr, destszExpr, false);
+        final Expression invalidCount = expressions.makeOr(countGtMax, countGtdestszExpr);
+        final Expression overlap = expressions.makeAnd(
+                expressions.makeGT(expressions.makeAdd(src, countExpr), dest, false),
+                expressions.makeGT(expressions.makeAdd(dest, countExpr), src, false));
+
+        final List<Event> replacement = new ArrayList<>();
+        
+        Label check1 = EventFactory.newLabel("__memcpy_s_check_1");
+        Label check2 = EventFactory.newLabel("__memcpy_s_check_2");
+        Label success = EventFactory.newLabel("__memcpy_s_success");
+        Label end = EventFactory.newLabel("__memcpy_s_end");
+
+        // dest == NULL or destsz > RSIZE_MAX ----> return error > 0
+        final Expression c1 = expressions.makeOr(destIsNull, invalidDestsz);
+        CondJump skipE1 = EventFactory.newJump(expressions.makeNot(c1), check2);
+        CondJump skipRest1 = EventFactory.newGoto(end);
+        Local retError1 = EventFactory.newLocal(resultRegister, expressions.makeOne((IntegerType)resultRegister.getType()));        
+        replacement.addAll(List.of(
+            check1,
+            skipE1,
+            retError1,
+            skipRest1));
+
+        // dest != NULL && destsz <= RSIZE_MAX && (src == NULL || count > destsz || overlap(src, dest)) 
+        // ----> return error > 0 and zero out [dest, dest+destsz)
+        // The first two are guaranteed by not matching c1
+        final Expression c2 = expressions.makeOr(srcIsNull, invalidCount);
+        final Expression c3 = expressions.makeOr(c2, overlap);
```

Comment:
I would combine these into one line. The intermediary `c2` is confusing. I intuitively thought that it designates a case like `c1` and `c3` but it was just an intermediary step to construct `c3`.
Also, you might want to rename those, cause `cx` could mean `condition x` or `case x` (if one thinks there are multiple cases to check for).
Maybe just have `c1 = condCase1` and `c3 = caseCond2 (getting rid of the intermediate c2)` and then just add a prefix to the comments: `Case 1: dest == null || ...`, `Case 2: dest != null && ...`, `Case 3: All inputs are valid`
I think that is easier to follow with one condition expression per case, and the fact that there are three cases to consider.